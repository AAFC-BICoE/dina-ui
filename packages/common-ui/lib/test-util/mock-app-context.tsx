import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
