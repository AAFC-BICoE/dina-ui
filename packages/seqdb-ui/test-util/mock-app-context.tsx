import {
  ApiClientContext,
  ApiClientContextI,
  createContextValue
} from "common-ui";
import { mount } from "enzyme";
import { SeqdbIntlProvider } from "../intl/seqdb-intl";

interface MockAppContextProviderProps {
  apiContext?: ApiClientContextI;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  apiContext,
  children
}: MockAppContextProviderProps) {
  return (
    <ApiClientContext.Provider value={apiContext || createContextValue()}>
      <SeqdbIntlProvider>{children}</SeqdbIntlProvider>
    </ApiClientContext.Provider>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
