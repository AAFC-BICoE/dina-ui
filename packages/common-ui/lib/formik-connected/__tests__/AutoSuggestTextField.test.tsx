import { KitsuResource } from "kitsu";
import AutoSuggest from "react-autosuggest";
import { mountWithAppContext } from "../../test-util/mock-app-context";
import { AutoSuggestTextField } from "../AutoSuggestTextField";
import { DinaForm } from "../DinaForm";

interface Person extends KitsuResource {
  name: string;
}

const PERSON_TEST_DATA_JSON_API = {
  data: [
    { name: "person1-json-api" },
    { name: "person2-json-api" },
    { name: "person3-json-api" }
  ]
};

const PERSON_TEST_DATA_ELASTICSEARCH = {
  data: {
    hits: [
      {
        source: {
          data: {
            id: "b7b24707-ad21-4957-a7ad-de423ba67094",
            type: "person",
            attributes: {
              name: "person1-elastic-search"
            }
          }
        }
      },
      {
        source: {
          data: {
            id: "a5f67abb-d9a5-4c21-8a5a-5339b71f8fd5",
            type: "person",
            attributes: {
              name: "person2-elastic-search"
            }
          }
        }
      },
      {
        source: {
          data: {
            id: "a4728a07-ad44-4086-a990-0d4ab7cf2105",
            type: "person",
            attributes: {
              name: "person3-elastic-search"
            }
          }
        }
      }
    ]
  }
};

// JSON API mock response.
const mockGet = jest.fn(async () => PERSON_TEST_DATA_JSON_API);

const mockGetFailure = jest.fn(async () => {
  throw new Error("Something went wrong!");
});

// Elastic Search mock response.
const mockGetAxios = jest.fn(async () => PERSON_TEST_DATA_ELASTICSEARCH);

// JSON API and Elastic Search mock responses.
const mockGetAll = jest.fn(async (path) => {
  if (path === "agent-api/person") {
    return PERSON_TEST_DATA_JSON_API;
  } else if (path === "search-api/search-ws/auto-complete") {
    return PERSON_TEST_DATA_ELASTICSEARCH;
  }
});

// JSON API only.
const apiContextJsonAPIOnly = {
  apiClient: {
    get: mockGet
  }
} as any;

// Elastic search only.
const apiContextElasticSearchOnly = {
  apiClient: {
    axios: {
      get: mockGetAxios
    }
  }
} as any;

// Both elastic search and JSON are working.
const apiContextBothProviders = {
  apiClient: {
    get: mockGetAll,
    axios: {
      get: mockGetAll
    }
  }
} as any;

// Elastic search works, but JSON API failure.
const apiContextJsonAPIFailure = {
  apiClient: {
    get: mockGetFailure,
    axios: {
      get: mockGetAxios
    }
  }
} as any;

// Json API works, but elastic search is not working.
const apiContextElasticSearchFailure = {
  apiClient: {
    get: mockGet,
    axios: {
      get: mockGetFailure
    }
  }
} as any;

const suggestionSelectedMock = jest.fn();

describe("AutoSuggestTextField", () => {
  // Clear the mocks between tests.
  beforeEach(jest.clearAllMocks);

  it("Snapshot test", async () => {
    const wrapper = mountWithAppContext(
      <DinaForm initialValues={{}}>
        <AutoSuggestTextField<Person>
          name="examplePersonNameField"
          jsonApiBackend={{
            query: (searchValue) => ({
              path: "agent-api/person",
              filter: {
                rsql: `name==*${searchValue}*`
              }
            }),
            option: (person) => person?.name
          }}
          onSuggestionSelected={suggestionSelectedMock}
          timeoutMs={0}
        />
      </DinaForm>,
      { apiContext: apiContextJsonAPIOnly }
    );

    // Snapshot test will check to ensure the layout does not change for unknown reasons.
    expect(wrapper.find(AutoSuggestTextField).debug()).toMatchSnapshot();

    // Simulate clicking an option, first need to provide a search.
    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "p" } });

    await new Promise(setImmediate);
    wrapper.update();

    // Select the first option in the list.
    wrapper.find(AutoSuggestTextField).find("Item").at(0).simulate("click");
    wrapper.update();

    expect(suggestionSelectedMock).toBeCalledTimes(1);
    expect(suggestionSelectedMock.mock.calls[0][0]).toEqual("person1-json-api");
  });

  it("JSON API only provided, results are fetched from JSON API", async () => {
    const wrapper = mountWithAppContext(
      <DinaForm initialValues={{}}>
        <AutoSuggestTextField<Person>
          name="examplePersonNameField"
          jsonApiBackend={{
            query: (searchValue) => ({
              path: "agent-api/person",
              filter: {
                rsql: `name==*${searchValue}*`
              }
            }),
            option: (person) => person?.name
          }}
          timeoutMs={0}
        />
      </DinaForm>,
      { apiContext: apiContextJsonAPIOnly }
    );

    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "p" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGet).lastCalledWith("agent-api/person", {
      filter: { rsql: "name==*p*" },
      sort: "-createdOn"
    });

    expect(mockGet).toHaveBeenCalledTimes(1);

    expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
      "person1-json-api",
      "person2-json-api",
      "person3-json-api"
    ]);

    // Test to ensure number of API calls does is not excessive.
    wrapper.find("input").simulate("change", { target: { value: "pe" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGet).toHaveBeenCalledTimes(2);

    // Try an empty search, since blankSearchProvider is not supplied it should not do any requests.
    wrapper.find("input").simulate("change", { target: { value: "" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGet).toHaveBeenCalledTimes(2);
  });

  it("Elastic Search only provided, results are fetched from elastic search.", async () => {
    const wrapper = mountWithAppContext(
      <DinaForm initialValues={{}}>
        <AutoSuggestTextField<Person>
          name="examplePersonNameField"
          elasticSearchBackend={{
            indexName: "dina_agent_index",
            searchField: "data.attributes.name",
            option: (person) => person?.name
          }}
          timeoutMs={0}
        />
      </DinaForm>,
      { apiContext: apiContextElasticSearchOnly }
    );

    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "p" } });

    await new Promise(setImmediate);
    wrapper.update();

    // Expected suggestions to appear.
    expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
      "person1-elastic-search",
      "person2-elastic-search",
      "person3-elastic-search"
    ]);

    // Take a snapshot of the suggestion popup being displayed in the UI.
    expect(
      wrapper.find(AutoSuggestTextField).find(".list-group").debug()
    ).toMatchSnapshot();

    // Remove focus from input, the suggestions should no longer be displayed.
    wrapper.find("input").simulate("blur");
    wrapper.update();
    expect(
      wrapper.find(AutoSuggestTextField).find(".list-group").exists()
    ).toBe(false);

    expect(mockGetAxios).lastCalledWith("search-api/search-ws/auto-complete", {
      params: {
        indexName: "dina_agent_index",
        autoCompleteField: "data.attributes.name",
        prefix: "p",
        additionalField: undefined,
        group: undefined
      }
    });

    expect(mockGetAxios).toHaveBeenCalledTimes(1);

    // Test to ensure number of API calls does is not excessive.
    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "pe" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGetAxios).toHaveBeenCalledTimes(2);

    // Try an empty search, since blankSearchProvider is not supplied it should not do any requests.
    wrapper.find("input").simulate("change", { target: { value: "" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGetAxios).toHaveBeenCalledTimes(2);
  });

  it("Custom options only provided, results come from custom options.", async () => {
    const wrapper = mountWithAppContext(
      <DinaForm initialValues={{}}>
        <AutoSuggestTextField<Person>
          name="examplePersonNameField"
          customOptions={(value) => [
            "suggestion-1",
            "suggestion-2",
            "suggestion-" + value
          ]}
          timeoutMs={0}
        />
      </DinaForm>,
      { apiContext: apiContextBothProviders }
    );

    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "3" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
      "suggestion-1",
      "suggestion-2",
      "suggestion-3"
    ]);

    expect(mockGetAll).toHaveBeenCalledTimes(0);
  });

  it("Both backend providers are used, preferred backend is used.", async () => {
    const wrapper = mountWithAppContext(
      <DinaForm initialValues={{}}>
        <AutoSuggestTextField<Person>
          name="examplePersonNameField"
          elasticSearchBackend={{
            indexName: "dina_agent_index",
            searchField: "data.attributes.name",
            option: (person) => person?.name
          }}
          jsonApiBackend={{
            query: (searchValue) => ({
              path: "agent-api/person",
              filter: {
                rsql: `name==*${searchValue}*`
              }
            }),
            option: (person) => person?.name
          }}
          preferredBackend={"json-api"}
          timeoutMs={0}
        />
      </DinaForm>,
      { apiContext: apiContextBothProviders }
    );

    wrapper.find("input").simulate("focus");
    wrapper.find("input").simulate("change", { target: { value: "p" } });

    await new Promise(setImmediate);
    wrapper.update();

    expect(mockGetAll).lastCalledWith("agent-api/person", {
      filter: { rsql: "name==*p*" },
      sort: "-createdOn"
    });

    expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
      "person1-json-api",
      "person2-json-api",
      "person3-json-api"
    ]);

    // Should not have any calls to elastic search.
    expect(mockGetAll.mock.calls).toEqual([
      [
        "agent-api/person",
        {
          filter: { rsql: "name==*p*" },
          sort: "-createdOn"
        }
      ]
    ]);
  });

  describe("Both backend providers are supplied, test what happens when a backend fails", () => {
    it("JSON-API is preferred, but fails, elastic search should be used instead.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <AutoSuggestTextField<Person>
            name="examplePersonNameField"
            elasticSearchBackend={{
              indexName: "dina_agent_index",
              searchField: "data.attributes.name",
              option: (person) => person?.name
            }}
            jsonApiBackend={{
              query: (searchValue) => ({
                path: "agent-api/person",
                filter: {
                  rsql: `name==*${searchValue}*`
                }
              }),
              option: (person) => person?.name
            }}
            preferredBackend={"json-api"}
            timeoutMs={0}
          />
        </DinaForm>,
        { apiContext: apiContextJsonAPIFailure }
      );

      wrapper.find("input").simulate("focus");
      wrapper.find("input").simulate("change", { target: { value: "p" } });

      await new Promise(setImmediate);
      wrapper.update();

      // JSON API should be tried first but fails.
      expect(mockGetFailure).toHaveBeenCalledTimes(1);
      expect(mockGetAxios).toHaveBeenCalledTimes(1);

      expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
        "person1-elastic-search",
        "person2-elastic-search",
        "person3-elastic-search"
      ]);

      wrapper.find("input").simulate("change", { target: { value: "pe" } });

      await new Promise(setImmediate);
      wrapper.update();

      // JSON API should not be called again at this point. Already failed and should have switched to
      // elastic search.
      expect(mockGetFailure).toHaveBeenCalledTimes(1);
      expect(mockGetAxios).toHaveBeenCalledTimes(2);
    });

    it("Elastic-search is preferred, but fails, JSON-API should be used instead.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <AutoSuggestTextField<Person>
            name="examplePersonNameField"
            elasticSearchBackend={{
              indexName: "dina_agent_index",
              searchField: "data.attributes.name",
              option: (person) => person?.name
            }}
            jsonApiBackend={{
              query: (searchValue) => ({
                path: "agent-api/person",
                filter: {
                  rsql: `name==*${searchValue}*`
                }
              }),
              option: (person) => person?.name
            }}
            preferredBackend={"elastic-search"}
            timeoutMs={0}
          />
        </DinaForm>,
        { apiContext: apiContextElasticSearchFailure }
      );

      wrapper.find("input").simulate("focus");
      wrapper.find("input").simulate("change", { target: { value: "p" } });

      await new Promise(setImmediate);
      wrapper.update();
      await new Promise(setImmediate);
      wrapper.update();

      // Elastic Search should be tried first, but fails.
      expect(mockGetFailure).toHaveBeenCalledTimes(1);
      expect(mockGet).toHaveBeenCalledTimes(1);

      expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
        "person1-json-api",
        "person2-json-api",
        "person3-json-api"
      ]);

      wrapper.find("input").simulate("change", { target: { value: "pe" } });

      await new Promise(setImmediate);
      wrapper.update();

      // JSON API should not be called again at this point. Already failed and should have switched to
      // json api.
      expect(mockGetFailure).toHaveBeenCalledTimes(1);
      expect(mockGet).toHaveBeenCalledTimes(2);
    });
  });

  describe("Blank search provider tests", () => {
    it("Blank search provider not supplied, options do not appear when empty search.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <AutoSuggestTextField<Person>
            name="examplePersonNameField"
            jsonApiBackend={{
              query: (searchValue) => ({
                path: "agent-api/person",
                filter: {
                  rsql: `name==*${searchValue}*`
                }
              }),
              option: (person) => person?.name
            }}
            timeoutMs={0}
          />
        </DinaForm>,
        { apiContext: apiContextJsonAPIOnly }
      );

      wrapper.find("input").simulate("focus");

      // Api should not be requested when search is empty since no blank search provider is supplied.
      expect(mockGet).toHaveBeenCalledTimes(0);
    });

    it("Blank search provider supplied, options should appear when empty search is focused.", async () => {
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <AutoSuggestTextField<Person>
            name="examplePersonNameField"
            jsonApiBackend={{
              query: (searchValue) => ({
                path: "agent-api/person",
                filter: {
                  rsql: `name==*${searchValue}*`
                }
              }),
              option: (person) => person?.name
            }}
            blankSearchBackend={"json-api"}
            timeoutMs={0}
          />
        </DinaForm>,
        { apiContext: apiContextJsonAPIOnly }
      );

      wrapper.find("input").simulate("focus");

      await new Promise(setImmediate);
      wrapper.update();

      // Expected suggestions to appear.
      expect(wrapper.find(AutoSuggest).prop("suggestions")).toEqual([
        "person1-json-api",
        "person2-json-api",
        "person3-json-api"
      ]);

      // Take a snapshot of the suggestion popup being displayed in the UI.
      expect(
        wrapper.find(AutoSuggestTextField).find(".list-group").debug()
      ).toMatchSnapshot();
    });
  });
});
