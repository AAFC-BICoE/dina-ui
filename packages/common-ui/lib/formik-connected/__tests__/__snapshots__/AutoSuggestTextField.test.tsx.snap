// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoSuggestTextField Blank search provider tests Blank search provider supplied, options should appear when empty search is focused. 1`] = `
"<ul role=\\"listbox\\" className=\\"list-group\\">
  <Item id=\\"react-autowhatever-examplePersonNameField--item-0\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={0} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={0} item=\\"person1-json-api\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-0\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={0} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person1-json-api
      </div>
    </li>
  </Item>
  <Item id=\\"react-autowhatever-examplePersonNameField--item-1\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={1} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={1} item=\\"person2-json-api\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-1\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={1} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person2-json-api
      </div>
    </li>
  </Item>
  <Item id=\\"react-autowhatever-examplePersonNameField--item-2\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={2} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={2} item=\\"person3-json-api\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-2\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={2} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person3-json-api
      </div>
    </li>
  </Item>
</ul>"
`;

exports[`AutoSuggestTextField Elastic Search only provided, results are fetched from elastic search. 1`] = `
"<ul role=\\"listbox\\" className=\\"list-group\\">
  <Item id=\\"react-autowhatever-examplePersonNameField--item-0\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={0} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={0} item=\\"person1-elastic-search\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-0\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={0} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person1-elastic-search
      </div>
    </li>
  </Item>
  <Item id=\\"react-autowhatever-examplePersonNameField--item-1\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={1} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={1} item=\\"person2-elastic-search\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-1\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={1} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person2-elastic-search
      </div>
    </li>
  </Item>
  <Item id=\\"react-autowhatever-examplePersonNameField--item-2\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={2} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]} sectionIndex={{...}} isHighlighted={false} itemIndex={2} item=\\"person3-elastic-search\\" renderItem={[Function: renderSuggestion]} renderItemData={{...}}>
    <li role=\\"option\\" id=\\"react-autowhatever-examplePersonNameField--item-2\\" aria-selected={false} className=\\"list-group-item\\" data-section-index={{...}} data-suggestion-index={2} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onMouseDown={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onClick={[Function (anonymous)]}>
      <div>
        person3-elastic-search
      </div>
    </li>
  </Item>
</ul>"
`;

exports[`AutoSuggestTextField Snapshot test 1`] = `
"<AutoSuggestTextField name=\\"examplePersonNameField\\" jsonApiBackend={{...}} onSuggestionSelected={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} timeoutMs={0}>
  <TextField name=\\"examplePersonNameField\\" customInput={[Function: customInput]}>
    <FieldWrapper name=\\"examplePersonNameField\\">
      <FieldSpy fieldName=\\"examplePersonNameField\\" validate={[undefined]}>
        <FormikConnect(FastFieldInner) name=\\"examplePersonNameField\\" shouldUpdate={[Function: shouldRender]} validate={[undefined]}>
          <FastFieldInner name=\\"examplePersonNameField\\" shouldUpdate={[Function: shouldRender]} validate={[undefined]} formik={{...}}>
            <FieldSpyInternal fastFieldProps={{...}} fieldName=\\"examplePersonNameField\\">
              <LabelWrapper fieldWrapperProps={{...}} fieldSpyProps={{...}}>
                <div className=\\"\\">
                  <label className=\\"examplePersonNameField-field mb-2 w-100 mb-3\\" htmlFor={[undefined]}>
                    <div className=\\"field-label mb-2\\">
                      <strong>
                        <FieldHeader name=\\"examplePersonNameField\\" customName={[undefined]} tooltipOverride={[undefined]} tooltipImage={[undefined]} tooltipImageAlt={[undefined]} tooltipLink={[undefined]} tooltipLinkText={[undefined]}>
                          <div className=\\"examplePersonNameField-field-header\\">
                            Example Person Name Field
                          </div>
                        </FieldHeader>
                      </strong>
                    </div>
                    <div className={[undefined]} style={{...}}>
                      <FormikConnectedField fieldSpyProps={{...}} fieldWrapperProps={{...}}>
                        <ErrorBoundary FallbackComponent={[Function: FallbackComponent]}>
                          <AutoSuggestTextFieldInternal elasticSearchBackend={[undefined]} jsonApiBackend={{...}} preferredBackend={[undefined]} blankSearchBackend={[undefined]} customOptions={[undefined]} placeholder=\\"Type here to search.\\" className=\\"form-control\\" onChange={[Function: onChange]} value=\\"\\" readOnly={[undefined]} onKeyDown={[Function: onKeyDown]} autoComplete=\\"none\\" onSuggestionSelected={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} id=\\"examplePersonNameField\\" timeoutMs={0}>
                            <style>
                              
                                      .autosuggest-container {
                                        position: relative;
                                      }
                                      .autosuggest .suggestions-container {
                                        display: none;
                                      }
                                      .autosuggest .suggestions-container-open {
                                        display: block;
                                        position: absolute;
                                        width: 100%;
                                        z-index: 20;
                                      }
                                      .autosuggest .suggestion-highlighted { 
                                        background-color: #ddd;
                                        cursor: pointer;
                                      }
                                    
                            </style>
                            <div className=\\"autosuggest\\">
                              <Autosuggest id=\\"examplePersonNameField\\" suggestions={{...}} getSuggestionValue={[Function: getSuggestionValue]} onSuggestionsFetchRequested={[Function: onSuggestionsFetchRequested]} onSuggestionSelected={[Function: onSuggestionSelected]} onSuggestionsClearRequested={[Function: onSuggestionsClearRequested]} renderSuggestion={[Function: renderSuggestion]} shouldRenderSuggestions={[Function: defaultShouldRenderSuggestions]} inputProps={{...}} theme={{...}} renderSuggestionsContainer={[Function: defaultRenderSuggestionsContainer]} alwaysRenderSuggestions={false} multiSection={false} shouldKeepSuggestionsOnSelect={[Function: shouldKeepSuggestionsOnSelect]} focusInputOnSuggestionClick={true} highlightFirstSuggestion={false} containerProps={{...}}>
                                <Autowhatever multiSection={false} items={{...}} renderInputComponent={[Function: defaultRenderInputComponent]} renderItemsContainer={[Function (anonymous)]} renderItem={[Function: renderSuggestion]} renderItemData={{...}} renderSectionTitle={[Function: renderSectionTitle]} getSectionItems={[Function: getSectionItems]} highlightedSectionIndex={{...}} highlightedItemIndex={{...}} containerProps={{...}} inputProps={{...}} itemProps={[Function (anonymous)]} theme={{...}} id=\\"examplePersonNameField\\">
                                  <div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"react-autowhatever-examplePersonNameField\\" aria-expanded={false} className=\\"autosuggest-container\\">
                                    <input type=\\"text\\" value=\\"\\" autoComplete=\\"none\\" aria-autocomplete=\\"list\\" aria-controls=\\"react-autowhatever-examplePersonNameField\\" aria-activedescendant={{...}} style={{...}} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} placeholder=\\"Type here to search.\\" className=\\"form-control\\" onChange={[Function: onChange]} readOnly={[undefined]} onKeyDown={[Function (anonymous)]} />
                                    <div id=\\"react-autowhatever-examplePersonNameField\\" role=\\"listbox\\" className=\\"suggestions-container\\" />
                                  </div>
                                </Autowhatever>
                              </Autosuggest>
                            </div>
                          </AutoSuggestTextFieldInternal>
                        </ErrorBoundary>
                      </FormikConnectedField>
                    </div>
                  </label>
                </div>
              </LabelWrapper>
            </FieldSpyInternal>
          </FastFieldInner>
        </FormikConnect(FastFieldInner)>
      </FieldSpy>
    </FieldWrapper>
  </TextField>
</AutoSuggestTextField>"
`;
